# Scripts used to Produce Manuscript Results and Figures

This directory contains a partial collection of scripts used to generate the results of the paper (see below), mostly with relation to motif discovery and analysis. 

## Sub-directories 
- `motif_discovery/`: scripts used to prepare runs of motif discovery software (Weeder, Seeder and MEME)
- `result_parsing/`: scripts used to parse the results of mootif discovery software and transfrom them into TAMO format
- `motif_analysis/`: scripts automating production of figures, clustering and general analysis of discovered motifs
- `promoter_database/`: scritps used to create a small SQLite database with location of S. tuberosum promoters in the genome

## Software versions
- Seeder v 0.01 (Fauteux, et al. 2008)
- Weeder v 2.0 (Zambelli, et al. 2014)
- MEME v 4.10.0 (Bailey et al. 2009)
- BioPython v 1.61 (Cock et al. 2009)
- Weblogo v 2.8 (Crooks et al. 2004)
- TAMO (Gordon, et al. 2005)

## Reference
Gálvez J.H., Tai H.H., Lagüe M., Zebarth B.J., Strömvik M. V. (2016) The nitrogen responsive transcriptome in potato (Solanum tuberosum L.) reveals significant gene regulatory motifs. Scientific Reports 6:26090.

Author: Hector Galvez